notebooks/search/tokenization.ipynb

```diff
@@ -0,0 +1,322 @@
+    "    return [whitespace_tokens, bert_tokens, e5_tokens, f\"{text[:80]}...\"]\n",
+    "\n",
+    "counts = [count_tokens(movie[\"plot\"]) for movie in movies]\n",
+    "\n",
+    "print(tabulate(sorted(counts), [\"whitespace\", \"BERT\", \"E5\", \"text\"]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Notice that both the BERT the E5 tokenizers yields more tokens in every example, in some cases even twice as many. Why is that? Let's look at an example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The lives of two mob hitmen, a boxer, a gangster and his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.\n",
+      "\n",
+      "['[CLS]', 'the', 'lives', 'of', 'two', 'mob', 'hit', '##men', ',', 'a', 'boxer', ',', 'a', 'gangster', 'and', 'his', 'wife', ',', 'and', 'a', 'pair', 'of', 'diner', 'bandits', 'inter', '##t', '##wine', 'in', 'four', 'tales', 'of', 'violence', 'and', 'redemption', '.', '[SEP]']\n"
+     ]
+    }
+   ],
+   "source": [
+    "example_movie = movies[0][\"plot\"]\n",
+    "print(example_movie)\n",
+    "print()\n",
+    "\n",
+    "movie_tokens = bert_tokenizer.encode(example_movie)\n",
+    "print(str([bert_tokenizer.decode([t]) for t in movie_tokens]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can observe:\n",
+    "- There are special tokens `[CLS]` and `[SEP]` to model the the beginning and end of the text. These two extra tokens will become relevant below.\n",
+    "- Punctuations are their own tokens.\n",
+    "- Compounds words are split into two tokens, for example `hitmen` becomes `hit` and `##men`.\n",
+    "\n",
+    "Given this behavior, it is easy to see how longer tests yield lots of tokens and can quickly get beyond the 512 tokens limitation mentioned above."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Handling long texts\n",
+    "\n",
+    "We saw how to count the number of tokens using the tokenizers from different models. ELSER uses the BERT tokenizer, so when using `.elser_model_2` it internally splits the text with this method.\n",
+    "\n",
+    "Currently there is a limitation that [only the first 512 tokens are considered](https://www.elastic.co/guide/en/machine-learning/8.12/ml-nlp-limitations.html#ml-nlp-elser-v1-limit-512). To work around this, we can first split the input text into chunks of 512 tokens and feed the chunks to Elasticsearch separately. Actually, we need to use a limit of 510 to leave space for the two special tokens (`[CLS]` and `[SEP]`) that we saw.\n",
+    "\n",
+    "Furthermore, it is best practice to make the chunks overlap (**TODO add reference**). With ELSER, we recommend 50% token overlap (i.e. a 256 token stride)."
```

Comment:
@joshdevins Do you have a reference that we can link here?

Reply:
I do not. Maybe @qherreros does? I don't think it's in public docs anywhere, just in our internal benchmarking.

Reply:
The best reference we have on this subject is in [internal documentation](https://docs.google.com/spreadsheets/d/1ojhOXu4rpl8dxKVqGUMfkYmt6vk5006CI7duFVimV5k/edit#gid=0).
I found this paper which is explicitly measuring improvements with overlaps, but it's in [encoder/decoder architecture](https://direct.mit.edu/tacl/article/doi/10.1162/tacl_a_00547/115346/Efficient-Long-Text-Understanding-with-Short-Text), fusing chunks in the decoder. I still think their conclusion can be applied to more than just FiD architecture.

Reply:
Alright, I went without a reference and this generic statement: _in practice we often see improved performance when using overlapping chunks_